crates/package_manager/src/add.rs

```diff
@@ -0,0 +1,96 @@
+use crate::Install;
+use derive_more::{Display, Error};
+use miette::Diagnostic;
+use pacquet_lockfile::Lockfile;
+use pacquet_npmrc::Npmrc;
+use pacquet_package_json::PackageJsonError;
+use pacquet_package_json::{DependencyGroup, PackageJson};
+use pacquet_registry::{PackageTag, PackageVersion};
+use pacquet_tarball::Cache;
+use reqwest::Client;
+
+/// This subroutine does everything `pacquet add` is supposed to do.
+#[must_use]
+pub struct Add<'a, ListDependencyGroups, DependencyGroupList>
+where
+    ListDependencyGroups: Fn() -> DependencyGroupList,
+    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
+{
+    /// Shared cache that store downloaded tarballs.
+    pub tarball_cache: &'a Cache,
+    /// HTTP client to make HTTP requests.
+    pub http_client: &'a Client,
+    /// Configuration read from `.npmrc`.
+    pub config: &'static Npmrc,
+    /// Data from the `package.json` file.
+    pub package_json: &'a mut PackageJson,
```

Comment:
In pnpm we use PackageManifest instead of PackageJson. I'd recommend using the same naming convention in this repository too.
```suggestion
    pub package_manifest: &'a mut PackageManifest,
```

Reply:
should be resolved by https://github.com/pnpm/pacquet/pull/143/commits/5f953bb3ad259155b51681b1d1690904b7b2c2e8 and https://github.com/pnpm/pacquet/pull/143/commits/a04e42ea3fe3ef8d823b92739c7aec43cfb813be